packages/babel-types/src/definitions/core.ts

```diff
@@ -1225,15 +1225,13 @@ defineType("VariableDeclarator", {
           return assertNodeType("LVal");
         }
 
-        const normal = assertNodeType(
+        const validator = assertNodeType(
           "Identifier",
           "ArrayPattern",
           "ObjectPattern",
         );
-        const without = assertNodeType("Identifier");
 
         return function (node: t.VariableDeclarator, key, val) {
-          const validator = node.init ? normal : without;
```

Comment:
This would have allowed invalid productions such as `var { a, b }`. Could you check if the `parent` (available as the first validate parameter) is a ForXStatement and the parent.left is node? Similar to line 1209.

Reply:
Good idea, but `node` has no `parent`:

```js
node {
  type: 'VariableDeclarator',
  id: {
    type: 'Placeholder',
    name: { type: 'Identifier', name: 'LHS', loc: [SourceLocation] },
    expectedNode: 'Pattern',
    loc: undefined
  },
  init: { type: 'NumericLiteral', value: 7 },
  loc: undefined
}
```

Reply:
Same with `validate` first argument: it is `undefined`:

```js
      validate: (function (a) {
        if (!process.env.BABEL_TYPES_8_BREAKING) {
          return assertNodeType("LVal");
        }

        const normal = assertNodeType(
          "Identifier",
          "ArrayPattern",
          "ObjectPattern",
        );
        const without = assertNodeType("Identifier");

        return function (node: t.VariableDeclarator, key, val) {
          console.log('node', node, 'key', key, 'value', val, 'a', a);
```

shows:

```js
node {
  type: 'VariableDeclarator',
  id: {
    type: 'Identifier',
    name: '$$BABEL_TPL$0',
    typeAnnotation: {
      type: 'TSTypeAnnotation',
      typeAnnotation: [Object],
      loc: undefined
    },
    loc: undefined
  },
  init: { type: 'Identifier', name: 'x', loc: undefined },
  loc: undefined
} key id value { type: 'ObjectPattern', properties: [] } a undefined
```

Reply:
@JLHwung just fixed, but as I see this error is already handled in `template` by parser